src/url.cpp

```diff
@@ -122,16 +122,37 @@ namespace Url
             if (index != std::string::npos)
             {
                 query_ = path_.substr(index + 1);
+                remove_repeats(query_, '?');
+                remove_repeats(query_, '&');
                 path_.resize(index);
             }
 
             index = path_.find(';');
             if (index != std::string::npos)
             {
                 params_ = path_.substr(index + 1);
+                remove_repeats(params_, ';');
                 path_.resize(index);
             }
         }
     }
 
+    void Url::remove_repeats(std::string& str, const char chr)
+    {
+        size_t dest = 0;
+        bool seen = true;
+        for (size_t src = 0; src < str.length(); ++src)
+        {
+            if (!seen || (str[src] != chr))
+            {
+                str[dest] = str[src];
```

Comment:
A very rough benchmark (take it a grain of salt):

```
Always copy, even if dest == src
================================
    
    ./bench "http://moz.com/????????????this&is&&&&&&&&an&&&&&&&example&&&&&with&&&lots"

        654.89 ms average of 5 runs of 1M parses

    ./bench "http://moz.com/?this&is&an&example&with&none"

        540.90 ms average of 5 runs of 1M parses

Check dest == src
=================

    ./bench "http://moz.com/????????????this&is&&&&&&&&an&&&&&&&example&&&&&with&&&lots"

        675.96 ms average of 5 runs of 1M parses

    ./bench "http://moz.com/?this&is&an&example&with&none"

        543.31 ms average of 5 runs of 1M parses
```

It seems that checking may be slightly slower, and is unlikely to be faster (even in the most favorable case)